// ngrx-slice/esm2022/lib/create-slice.mjs
import { createFeatureSelector } from '@ngrx/store';
import { createSliceActions } from './create-slice-actions';
import { createSliceReducer } from './create-slice-reducer';
import { createSliceSelectors } from './create-slice-selectors';
import { classify } from './utils';
function defaultSliceActionNameGetter(featureName, actionName) {
    return `[${classify(featureName)}] ${actionName}`;
}
export function noopReducer() {
    return (() => { });
}
export function createSlice({ name, initialState, reducers, extraReducers, reducersToActions, sliceActionNameGetter = defaultSliceActionNameGetter, }) {
    const featureSelector = createFeatureSelector(name);
    const nestedSelectors = createSliceSelectors(initialState, featureSelector);
    const actions = createSliceActions(name, sliceActionNameGetter, reducers, reducersToActions);
    const reducer = createSliceReducer(initialState, actions, reducers, extraReducers);
    return {
        name,
        reducer,
        actions,
        selectors: {
            [`select${classify(name)}State`]: featureSelector,
            ...nestedSelectors,
        },
    };
}
export function createNamespacedSlice({ name, initialState, reducers, extraReducers, reducersToActions, sliceActionNameGetter = defaultSliceActionNameGetter, }) {
    const { name: sliceName, reducer, selectors, actions, } = createSlice({
        name,
        initialState,
        reducers,
        extraReducers,
        reducersToActions,
        sliceActionNameGetter,
    });
    const classifiedName = classify(name);
    return {
        [`${classifiedName}Feature`]: {
            name: sliceName,
            reducer,
        },
        [`${classifiedName}Actions`]: actions,
        [`${classifiedName}Selectors`]: selectors,
    };
}
